### PR TITLE
[Optimization] Collapse opacity filters where possible into primitive.

### PR DESCRIPTION
**Pull request by glennw**
In many cases, an opacity filter is used to make a rectangle or
image partially transparent, and only contains a single primitive.

In these cases, we can skip drawing the primitive to an off-screen
surface and applying the opacity while blitting that target to the
parent surface.

Instead, we collapse the opacity value(s) into the color of the
primitive, and skip the intermediate surface altogether, drawing
the primitive directly into the parent surface.

This is a very common use case on real world sites, so it has a
significant performance win in many sites.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/2680)

### CODE REVIEW

**Comment by glennw** [comment]
r? @kvark or @mrobinson 

Try run is pending, but looks good so far:
https://treeherder.mozilla.org/#/jobs?repo=try&revision=96a6494b8062e57dcdbf8d5084d34e2860a638f2

**Comment by glennw** [comment]
Hmmm, a couple of Windows-only failures that look to be related - they are animated opacity.

Although, there are intermittent bugs filed for both of these that appear to have happened on other builds in the last seven days, so maybe they are actually unrelated?

Any thoughts @staktrace on whether these need to be investigated? (I can't repro them on Linux locally or with CI).

**Comment by staktrace** [comment]
@glennw I think the intermittents are pre-existing, not related to your patch. Feel free to land.

File: webrender/src/prim_store.rs [hunk 1]
```diff
@@ -1166,6 +1220,118 @@ impl PrimitiveStore {
...
+                    let brush = &mut self.cpu_brushes[prim_metadata.cpu_prim_index.0];
+
+                    // By this point, we know we should only have found a primitive
+                    // that supports opacity collapse.
+                    match brush.kind {
+                        BrushKind::Solid { ref mut opacity_binding, .. } |
+                        BrushKind::Image { ref mut opacity_binding, .. } => {
+                            opacity_binding.push(binding);
+                        }
+                        BrushKind::Clear { .. } |
```
Comment (mrobinson) [inline webrender/src/prim_store.rs]: Another good candidate for the wildcard match?

File: webrender/src/prim_store.rs [hunk 2]
```diff
@@ -1166,6 +1220,118 @@ impl PrimitiveStore {
...
+    // that can be the target for collapsing parent opacity filters into.
+    fn get_opacity_collapse_prim(
+        &self,
+        pic_index: PictureIndex,
+    ) -> Option<PrimitiveIndex> {
+        let pic = &self.pictures[pic_index.0];
+
+        // We can only collapse opacity if there is a single primitive, otherwise
+        // the opacity needs to be applied to the primitives as a group.
+        if pic.runs.len() == 1 {
```
[the remark this answers is not in the document]
Comment (glennw) [reply webrender/src/prim_store.rs]: Fixed

File: webrender/src/prim_store.rs [hunk 3]
```diff
@@ -1166,6 +1220,118 @@ impl PrimitiveStore {
...
+                                if pic.composite_mode.is_none() {
+                                    return self.get_opacity_collapse_prim(pic_index);
+                                }
+                            }
+                            // If we find a single rect or image, we can use that
+                            // as the primitive to collapse the opacity into.
+                            BrushKind::Solid { .. } | BrushKind::Image { .. } => {
+                                return Some(run.base_prim_index)
+                            }
+                            BrushKind::YuvImage { .. } |
```
[the remark this answers is not in the document]
Comment (glennw) [reply webrender/src/prim_store.rs]: Totally subjective - but I feel like wildcards in matches are a bit of an anti-pattern where they can be avoided. The downside of the current code is that it's more verbose, but the upside is that whenever we port a new primitive type to be a brush, we get a compiler error everywhere that the brush kind is matched on, which makes it really simple to evaluate each call site to see how the new brush type should be handled at this location.

**Comment by glennw** [comment]
Rebased

File: webrender/src/prim_store.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (mrobinson) [reply webrender/src/prim_store.rs]: Okay. I feel like that's a pretty reasonable stance. I'll try to use them less when I am working on WebRender.

**Pull request merged**